src/hotspot/share/gc/g1/g1EvacFailure.cpp

```diff
@@ -259,5 +259,8 @@ G1ParRemoveSelfForwardPtrsTask::G1ParRemoveSelfForwardPtrsTask(G1RedirtyCardsQue
 void G1ParRemoveSelfForwardPtrsTask::work(uint worker_id) {
   RemoveSelfForwardPtrHRClosure rsfp_cl(_rdcqs, worker_id);
 
-  _g1h->collection_set_iterate_increment_from(&rsfp_cl, &_hrclaimer, worker_id);
+  // We need to check all regions whether they need self forward removals, not only
+  // the last collection set increment. Reference processing may copy over and fail
+  // evacuation in any region in the collection set.
```

Comment:
I think your explanation in the summary is very clear and it would be nice to capture it in the comment as well. Maybe something like: 
```suggestion
  // We need to check all regions whether they need self forward removals, not only
  // the last collection set increment. The reason is that reference processing (e.g. 
  // finalizers) can make it necessary to resurrect an otherwise unreachable object at
  // the very end of the collection. The resurrected object might be located in a region
  // evacuated in an earlier increment, but copying it at the end of the collection can 
  // trigger an evacuation failure.
```